examples/ota-provider-app/linux/main.cpp

```diff
@@ -62,6 +62,43 @@ constexpr chip::System::Clock::Timeout kBdxPollFreq = chip::System::Clock::Milli
 OTAProviderExample::queryImageBehaviorType gQueryImageBehavior = OTAProviderExample::kRespondWithUpdateAvailable;
 uint32_t gDelayedActionTimeSec                                 = 0;
 const char * gOtaFilepath                                      = nullptr;
+uint32_t otaTransferInProgress                                 = false;
+
+void OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset);
+void OnTransferComplete(void * context);
+void OnTransferFailed(void * context);
+
+chip::Callback::Callback<OnBdxBlockQuery> mOnBlockQuery(OnBlockQuery, nullptr);
+chip::Callback::Callback<OnBdxTransferComplete> mOnTransferComplete(OnTransferComplete, nullptr);
+chip::Callback::Callback<OnBdxTransferFailed> mOnTransferFailed(OnTransferFailed, nullptr);
+
+void OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset)
+{
+    std::ifstream otaFile(gOtaFilepath, std::ifstream::in);
+    VerifyOrReturn(otaFile.good(), ChipLogError(BDX, "%s: file read failed", __FUNCTION__));
+    otaFile.seekg(offset);
+    otaFile.read(reinterpret_cast<char *>(blockBuf->Start()), kMaxBdxBlockSize);
+    if ((otaFile.good() || otaFile.eof()) == false)
+    {
+        ChipLogError(BDX, "%s: file read failed", __FUNCTION__);
+        return;
+    }
+
+    size  = static_cast<size_t>(otaFile.gcount());
+    isEof = (size < kMaxBdxBlockSize || otaFile.peek() == EOF);
```

Comment:
This doesn't account for finite-length transfers. It only works if we're serving an entire file all the way to the end, but this won't always be the case.

Again, the `TransferSession` object inside `BdxOtaSender` has the length information that we need here. This is a limitation of this current callback design - the callbacks need information from `BdxOtaSender` but they're defined outside that scope.